sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/sqs/SqsIO.java

```diff
@@ -0,0 +1,272 @@
+
+      if (maxNumRecords() < Long.MAX_VALUE || maxReadTime() != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime()).withMaxNumRecords(maxNumRecords());
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+  }
+
+  /**
+   * A {@link PTransform} to send messages to SQS. See {@link SqsIO} for more information on usage
+   * and configuration.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<SendMessageRequest>, PDone> {
+
+    @Nullable
+    abstract SqsClientProvider getSqsClientProvider();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+
+      abstract Builder setSqsClientProvider(SqsClientProvider sqsClientProvider);
+
+      abstract Write build();
+    }
+
+    /**
+     * Allows to specify custom {@link SqsClientProvider}. {@link SqsClientProvider} creates new
+     * {@link SqsClient} which is later used for writing to a SqS queue.
+     */
+    public Write withSqsClientProvider(SqsClientProvider awsClientsProvider) {
+      return builder().setSqsClientProvider(awsClientsProvider).build();
+    }
+
+    /**
+     * Specify {@link software.amazon.awssdk.auth.credentials.AwsCredentialsProvider} and region to
+     * be used to write to SQS. If you need more sophisticated credential protocol, then you should
+     * look at {@link Write#withSqsClientProvider(SqsClientProvider)}.
+     */
+    public Write withSqsClientProvider(AwsCredentialsProvider credentialsProvider, String region) {
+      return withSqsClientProvider(credentialsProvider, region, null);
+    }
+
+    /**
+     * Specify {@link AwsCredentialsProvider} and region to be used to write to SQS. If you need
+     * more sophisticated credential protocol, then you should look at {@link
+     * Write#withSqsClientProvider(SqsClientProvider)}.
+     *
+     * <p>The {@code serviceEndpoint} sets an alternative service host. This is useful to execute
+     * the tests with Kinesis service emulator.
+     */
+    public Write withSqsClientProvider(
+        AwsCredentialsProvider credentialsProvider, String region, URI serviceEndpoint) {
+      return withSqsClientProvider(
+          new BasicSqsClientProvider(credentialsProvider, region, serviceEndpoint));
+    }
+
+    @Override
+    public PDone expand(PCollection<SendMessageRequest> input) {
+      input.apply(ParDo.of(new SqsWriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+  }
+
+  private static class SqsWriteFn extends DoFn<SendMessageRequest, Void> {
+    private final Write spec;
+    private transient SqsClient sqs;
+
+    SqsWriteFn(Write write) {
+      this.spec = write;
+    }
+
+    @Setup
+    public void setup() {
+      sqs = spec.getSqsClientProvider().getSqsClient();
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext processContext) throws Exception {
+      sqs.sendMessage(processContext.element());
```

Comment:
Does it send message synchronically? Would it make sense to batch the messages of the whole bundle and send them as a batch for better performance?

Reply:
Yes, that make send, but then our API has to take in SendMessageBatchRequestEntry, instead of SendMessageRequest. Now, there are three ways to deal with this:
1) Take in SendMessageBatchRequestEntry, instead of SendMessageRequest. This way will impact all of the existing users
2) Keep the existing api, which take in SendMessageRequest, and convert it to SendMessageBatchRequestEntry manually in our IO. Bascially, mapping each attribute of SendMessageRequest to SendMessageBatchRequest's
3) Add another API, called `writeBatch`, and this one takes in SendMessageBatchRequestEntry,  backward compatibility support
I prefer #3, since it makes a clear distinction, users know what they are doing, and nobody get "hurt" ;-) 
@aromanenko-dev let me know what you think?

Reply:
I'm ok to keep it as it was before for now. Perhaps, it would make sense to add TODO comment for the future about that.